### PR TITLE
Make _freq/freq/tz/_tz/dtype/_dtype/offset/_offset all inherit reliably

### PR DESCRIPTION
**Pull request by jbrockmendel**
There was an issue in rebasing #24024 in which index._freq didn't match index._eadata._freq.  This ensures that is never an issue by removing _freq/freq from the Index subclasses and making them properties aliasing the _eadata attrs.

To do this, we have to make _eadata not-a-property for the time being.

Also make the _tz --> _dtype transition in #24024, and fix a couple of places in DatetimeIndex where it sets _tz manually that would otherwise be missed.

Remove PeriodIndex.shift, since it now can use the DatetimeIndexOpsMixin version (also done in 24024)

### CODE REVIEW

File: pandas/core/arrays/datetimes.py [hunk 1]
```diff
@@ -210,7 +210,7 @@ class DatetimeArrayMixin(dtl.DatetimeLikeArrayMixin,
     # Constructors
 
     _attributes = ["freq", "tz"]
-    _tz = None
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimes.py]: Can you add type comments here? `  # type: Union[np.dtype, DatetimeTZDtype]`.

File: pandas/core/arrays/datetimelike.py [hunk 2]
```diff
@@ -1138,8 +1138,8 @@ def _time_shift(self, periods, freq=None):
                 freq = frequencies.to_offset(freq)
             offset = periods * freq
             result = self + offset
-            if hasattr(self, 'tz'):
-                result._tz = self.tz
+            if getattr(self, 'tz', None):
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimelike.py]: On #24024 at least, this is not necessary.

File: pandas/core/arrays/datetimes.py [hunk 3]
```diff
@@ -399,9 +404,12 @@ def dtype(self):
             If the values are tz-aware, then the ``DatetimeTZDtype``
             is returned.
         """
-        if self.tz is None:
-            return _NS_DTYPE
-        return DatetimeTZDtype('ns', self.tz)
+        return self._dtype
+
+    @property
```
Comment (TomAugspurger) [inline pandas/core/arrays/datetimes.py]: Hmm why was this necessary? Ideally everyone would reference `self.dtype.tz`

Comment (TomAugspurger) [reply pandas/core/arrays/datetimes.py]: I think if you update the `tz` property defend below to be 

```python
    @property
    def tz(self):
        """
        Return timezone, if any.

        Returns
        -------
        datetime.tzinfo, pytz.tzinfo.BaseTZInfo, dateutil.tz.tz.tzfile, or None
            Returns None when the array is tz-naive.
        """
        # GH 18595
        return getattr(self.dtype, 'tz', None)

    @tz.setter
    def tz(self, value):
        # GH 3746: Prevent localizing or converting the index by setting tz
        raise AttributeError("Cannot directly set timezone. Use tz_localize() "
                             "or tz_convert() as appropriate")
```

then things will hopefully work.

File: pandas/core/indexes/datetimes.py [hunk 4]
```diff
@@ -321,11 +321,10 @@ def _simple_new(cls, values, name=None, freq=None, tz=None, dtype=None):
 
         dtarr = DatetimeArray._simple_new(values, freq=freq, tz=tz)
         result = object.__new__(cls)
-        result._data = dtarr._data
-        result._freq = dtarr.freq
-        result._tz = dtarr.tz
+        result._eadata = dtarr
         result.name = name
         # For groupby perf. See note in indexes/base about _index_data
+        # TODO: make sure this is updated correctly if edited
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: This TODO is for _index_data? In theory that shouldn't happen, since DatetimeIndex is immutable.

File: pandas/core/indexes/datetimes.py [hunk 5]
```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
     # Wrapping DatetimeArray
 
     @property
-    def _eadata(self):
-        return DatetimeArray._simple_new(self._data,
-                                         tz=self.tz, freq=self.freq)
+    def _data(self):
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: I don't think that `_data` is a property anywhere else. Just an attribute that's set in `_simple_new`.

Actually... how does this even work? If you don't have a setter (which I don't see) then simple_new should fail.

File: pandas/core/indexes/datetimes.py [hunk 6]
```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
...
+        return self._eadata._data
+
+    @property
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: Can the private versions be removed?

File: pandas/core/indexes/datetimes.py [hunk 7]
```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
...
+    def _freq(self):
+        return self._eadata._freq
+
+    @_freq.setter
+    def _freq(self, value):
+        # Validation will be handled by _eadata setter
+        self._eadata._freq = value
+
+    @property
+    def freq(self):
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: This should reference `self._eadata.freq`

File: pandas/core/indexes/datetimes.py [hunk 8]
```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
...
+        return self._freq
+
+    @freq.setter
+    def freq(self, value):
+        # validation is handled by _eadata setter
+        self._eadata.freq = value
+
+    @property
+    def _tz(self):
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: Remove?

File: pandas/core/indexes/datetimes.py [hunk 9]
```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
...
+        return self._eadata._tz
+
+    @property
+    def tz(self):
+        # GH#18595
+        return self._tz
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: return `self._eadata.tz`

File: pandas/core/indexes/datetimes.py [hunk 10]
```diff
@@ -1129,9 +1129,43 @@ def slice_indexer(self, start=None, end=None, step=None, kind=None):
...
+
+    @tz.setter
+    def tz(self, value):
+        # GH 3746: Prevent localizing or converting the index by setting tz
+        raise AttributeError("Cannot directly set timezone. Use tz_localize() "
```
Comment (TomAugspurger) [inline pandas/core/indexes/datetimes.py]: I wonder if this can just do `self._eadata.tz = value`, which will raise the AttributeError for us? Then we don't duplicate this code.

File: pandas/core/indexes/datetimes.py (hunk 5 above)
Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: Ah I see you set eadata.

It doesn't really matter since we're removing it soon anyway, but I'd prefer that `_eadata` always reference `_data`. That makes keeps all the index classes consistent that `._data` is the actual array, not a property.

File: pandas/core/indexes/datetimes.py (hunk 7 above)
Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: Also, should be able to move this to DatetimeIndexOpsMixin.

**Comment by TomAugspurger** [comment]
While you're at it, could you add

1. Add `DatetimeIndexOpsMixin._ndarray_values` as

```python
@property
def _ndarray_values(self):
    return self._eadata._ndarray_values
``

2. Remove `PeriodIndex._ndarray_values`

File: pandas/core/indexes/datetimes.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: In _libs.reduction there is a line:

```
object.__setattr__(cached_ityp, '_index_data', islider.buf)
```

which makes me wary.  Is this just never relevant for DTI?

File: pandas/core/indexes/datetimes.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: > I don't think that _data is a property anywhere else. Just an attribute that's set in _simple_new.

That's right.  In this PR we set _eadata inside _simple_new and make _data a property returning _eadata._data.  The freq/tz passthrough doesn't work with _eadata as a property (as in master), so the only question is whether to _also_ set _data in _simple_new.  I chose to make it a property to prevent any shenanigans where they become untied.

Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: I think https://github.com/pandas-dev/pandas/pull/24024/files#diff-26a6d2ca7adfca586aabbb1c9dd8bf36R74 is what we want for eadata & freq (and if we can do it here, instead of that PR, then that's best).

File: pandas/core/indexes/datetimes.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: That's the intent of all this, to directly mutate the buffer in place. The `.reset` method undoes all this stuff. Nobody else should be messing with it.

File: pandas/core/indexes/datetimes.py [hunk 11]
```diff
@@ -532,9 +528,13 @@ def union_many(self, others):
             else:
                 tz = this.tz
                 this = Index.union(this, other)
-                if isinstance(this, DatetimeIndex):
-                    this._tz = timezones.tz_standardize(tz)
-
+                if isinstance(this, DatetimeIndex) and tz is not None:
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: this is what the ._tz setter used to do, so just make a method to do instead (as you are calling in multiple places), maybe ``_set_timezone``

File: pandas/core/indexes/timedeltas.py [hunk 12]
```diff
@@ -279,8 +280,25 @@ def _format_native_types(self, na_rep='NaT', date_format=None, **kwargs):
     # Wrapping TimedeltaArray
 
     @property
-    def _eadata(self):
-        return TimedeltaArray._simple_new(self._data, freq=self.freq)
+    def _data(self):
```
Comment (jreback) [inline pandas/core/indexes/timedeltas.py]: can't these just be in the DatetimeIndexOpsMixin?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/timedeltas.py (hunk 12 above)
Comment (jbrockmendel) [reply pandas/core/indexes/timedeltas.py]: We could, but would have to override for PeriodIndex.  Either way is fine by me.

File: pandas/core/indexes/datetimes.py (hunk 11 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: OK.  I'm kind of hoping this is all unnecessary, will take a look

File: pandas/core/indexes/timedeltas.py (hunk 12 above)
Comment (jreback) [reply pandas/core/indexes/timedeltas.py]: ok i c. either way is fine then.

File: pandas/core/indexes/datetimes.py (hunk 11 above)
Comment (jreback) [reply pandas/core/indexes/datetimes.py]: ok its fine for now.

Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: just pushed with a fix here: just set `self._eadata._dtype`, analogous to the previous method setting `self._tz`.

Although in all of the tests where this is hit, it already has the appropriate dtype, so this may be superfluous.

File: pandas/core/indexes/datetimes.py (hunk 4 above)
Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: @TomAugspurger I think all your other comments have been addressed, not sure about this one.  Should this TODO comment be removed?  Some other action taken?

File: pandas/core/indexes/datetimes.py (hunk 11 above)
Comment (jreback) [reply pandas/core/indexes/datetimes.py]: k

File: pandas/core/arrays/datetimes.py [hunk 13]
```diff
@@ -231,8 +231,13 @@ def _simple_new(cls, values, freq=None, tz=None):
         result = object.__new__(cls)
         result._data = values
         result._freq = freq
-        tz = timezones.maybe_get_tz(tz)
-        result._tz = timezones.tz_standardize(tz)
+        if tz is None:
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: is it cheap to detect when you need to call maybe_get_tz & tz_standarize? ideally these could be called in the DatetimeTZDtype constructor / alternatively could have a dedicate constructor, 

``DateteimeTZDtype.construct_from_tztype``, and will pave the way for a unified dtype as well.

**Review by jreback — APPROVED** [review]
just a small comment on future directions about dtypes.

**Comment by jreback** [comment]
ping on green.

**Comment by jbrockmendel** [comment]
ping

**Comment by jreback** [comment]
thanks!

File: pandas/core/indexes/datetimes.py (hunk 4 above)
Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: Yeah this todo isn't necessary AFAICT.

Also, I really don't think we should have inverted the relationship between the eadata and data attributes.

Comment (TomAugspurger) [reply pandas/core/indexes/datetimes.py]: Just because that change isn't going to last through #24024, so I think it was unnecessary.

Comment (jbrockmendel) [reply pandas/core/indexes/datetimes.py]: > Also, I really don't think we should have inverted the relationship between the eadata and data attributes. [...] so I think it was unnecessary.

We can't have both 1) _eadata being a property that depends on self.freq and 2) freq be a property that depends on _eadata.

Definitely agree that 24024 should return things to the old pattern.

File: pandas/core/arrays/datetimes.py [hunk 14]
```diff
@@ -411,10 +414,10 @@ def tz(self):
         Returns
         -------
         datetime.tzinfo, pytz.tzinfo.BaseTZInfo, dateutil.tz.tz.tzfile, or None
-             Returns None when the array is tz-naive.
+            Returns None when the array is tz-naive.
         """
         # GH 18595
-        return self._tz
+        return getattr(self._dtype, "tz", None)
```
[the remark this answers is not in the document]
Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: No, feel free to change
